clash-lib/src/Clash/Core/HasType.hs

```diff
@@ -0,0 +1,221 @@
+Maintainer  : QBayLogic B.V. <devops@qbaylogic.com>
+
+Utility class to extract type information from data which has a type.
+-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Core.HasType
+  ( HasType(..)
+  , InferType(..)
+  , applyTypeToArgs
+  , piResultTy
+  , piResultTys
+  ) where
+
+import qualified Data.Text as Text (isInfixOf)
+import Data.Text.Prettyprint.Doc (line)
+import GHC.Stack (HasCallStack)
+
+import Clash.Core.DataCon (DataCon(dcType))
+import Clash.Core.FreeVars
+import Clash.Core.Literal (Literal(..))
+import Clash.Core.Name (Name(nameOcc))
+import Clash.Core.Pretty
+import Clash.Core.Subst
+import Clash.Core.Term (Term(..), IsMultiPrim(..), PrimInfo(..), collectArgs)
+import Clash.Core.TyCon (TyConMap)
+import Clash.Core.Type
+import Clash.Core.TysPrim
+import Clash.Core.Var (Var(varType))
+import Clash.Core.VarEnv
+import Clash.Debug (debugIsOn)
+import Clash.Util (pprPanic)
+import qualified Clash.Util.Interpolate as I
+
+class HasType a where
+  coreTypeOf :: a -> Type
+
+instance HasType DataCon where
+  coreTypeOf = dcType
+
+instance HasType Literal where
+  coreTypeOf = \case
+    IntegerLiteral _ -> integerPrimTy
+    IntLiteral _ -> intPrimTy
+    WordLiteral _ -> wordPrimTy
+    StringLiteral _ -> stringPrimTy
+    FloatLiteral _ -> floatPrimTy
+    DoubleLiteral _ -> doublePrimTy
+    CharLiteral _ -> charPrimTy
+    Int64Literal _ -> int64PrimTy
+    Word64Literal _ -> word64PrimTy
+    NaturalLiteral _ -> naturalPrimTy
+    ByteArrayLiteral _ -> byteArrayPrimTy
+
+instance HasType PrimInfo where
+  coreTypeOf pr =
+    case primMultiResult pr of
+      SingleResult -> primType pr
+
+      -- See Note [MultiResult type] in Clash.Normalize.Transformations.MultiPrim
+      MultiResult
+        | let (primArgs, primResTy) = splitFunForallTy (primType pr)
+        , TyConApp tupTcNm tupArgs <- tyView primResTy
+        , Text.isInfixOf "GHC.Tuple.(" (nameOcc tupTcNm)
+        -> mkPolyFunTy primResTy (primArgs <> fmap Right tupArgs)
+
+        | otherwise
+        -> error "PrimInfo.coreTypeOf: MultiResult primitive without tuple type"
+
+instance HasType Type where
+  coreTypeOf = id
+
+instance HasType (Var a) where
+  coreTypeOf = varType
+
+class InferType a where
+  inferCoreTypeOf :: TyConMap -> a -> Type
+
+instance InferType Term where
+  inferCoreTypeOf tcm = go
+   where
+    go = \case
+      Var i -> coreTypeOf i
+      Data dc -> coreTypeOf dc
+      Literal l -> coreTypeOf l
+      Prim pr -> coreTypeOf pr
+      Lam i x -> mkFunTy (coreTypeOf i) (go x)
+      TyLam i x -> ForAllTy i (go x)
+
+      x@App{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      x@TyApp{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      Letrec _ x -> go x
+      Case _ _ [] -> undefinedTy
+      Case _ _ (x:_) -> go (snd x)
```

Comment:
why not just:
```
Case _ ty _ -> ty
```
like we did in `termType`?

Since we seem to have replaced all occurrences of `termType` by `inferCoreType`, traversing the alternative here probably makes the compiler slower.

Reply:
Also, since we eliminate casts in our GHC2Core pass, when patterns contains existentials, the RHS of the alternative might not even have the correct type. Like:
```
case (x :: Vec n a) of
  Cons m (n ~ m + 1) (y :: a) (ys :: m) -> Cons m CO y ys |> (Vec (m+1) a ~ Vec n a)
```
with the cast removed would return:
```
Vec (m + 1) a
```
instead of the correct type:
```
Vec n a
```

Reply:
> why not just:
>
> Case _ ty _ -> ty
> like we did in termType?

Isn't this wrong though? I thought this was the type of the scrutinee, not the result

> when patterns contains existentials, the RHS of the alternative might not even have the correct type

Yeah... that would be a problem. I guess if it shouldn't be changed back to `Case _ ty _ -> ty` then it needs a note